### ISSUE TITLE
Make test great again!

### ISSUE THREAD

**Issue by denismakogon** [fnproject/fn#741]
Okay, our tests are not developer-friendly.

 

- [ ]  datastore tests, they are not modular so can't run them with usual go test flexibility.
- [ ]  every package or test file can't be tested separately:
```bash
go test -v agent/agent_test.go 
# command-line-arguments
agent/agent_test.go:95:7: undefined: New
agent/agent_test.go:95:11: undefined: NewDirectDataAccess
agent/agent_test.go:115:3: undefined: WithWriter
agent/agent_test.go:116:3: undefined: FromRequest
agent/agent_test.go:238:7: undefined: New
agent/agent_test.go:238:11: undefined: NewDirectDataAccess
agent/agent_test.go:241:26: undefined: FromModel
agent/agent_test.go:309:7: undefined: New
agent/agent_test.go:309:11: undefined: NewDirectDataAccess
agent/agent_test.go:312:26: undefined: FromModel
agent/agent_test.go:312:26: too many errors
FAIL	command-line-arguments [build failed]
```
and so on, there might be more things to fix, but we should pay more attention to the tests we do!

**Comment by rdallman** [comment]
i would say the `server/` tests are the not friendly ones, they aren't flexible really since we have to parse the responses out. the datastore tests are really fine? they just only test the datastore, and there are other things to test. the integration tests we have are largely a duplication of the `server/` tests and seem to mostly exist to test swagger client generation (cli uses) but duplicate most of what `server` tests already do. 

agent tests exist since server tests are hard to work with. 

your command to run `go test -v agent/agent_test.go` is completely untenable to expect for a project of this size to be able to do. you're missing definitions since go test only builds that file. if you do `go test` in that directory it works fine, so think you can remove that particular nit. if you want to run particular tests go test has `-run=`. 

about the last thing we need is another test suite. if anything, we should kill some. this isn't a great issue since there isn't a concrete plan proposed

**Comment by denismakogon** [comment]
I tend to disagree and here's why. API tests live in Fn repo because they don't have their own home, initially, they were designed to be a validation test suite for any Fn deployments. So the nature of those tests really looks similar to `/server` tests, except those cases when server tests don't do the whole procedure of app/route deploying and use native HTTP client to talk to the test server.

I do agree, that i haven't proposed particular work plan, but would be nice to have a placeholder for thoughts like yours to come up with the exact action items.

**Comment by rdallman** [comment]
sure. but since we have a swagger api i kind of can't imagine why we need such a rigorous set of api level tests hanging off on the side, when we can test the more intricate stuff much more easily without putting HTTP in between if we design the code reasonably well. right now as far as being a dev, we have 2 tests where you have to parse json blobs out of http before checking anything. this removes all type checking possibilities, at a minimum, and is otherwise simply an unnecessary pain.
